Use mpark::variant on Cocoa platforms

achristensen07 opened the pull request:
#### badfefea5cf9b314c1aeaaae17244a1c9fbc9c34
<pre>
Use mpark::variant on Cocoa platforms
<a href="https://bugs.webkit.org/show_bug.cgi?id=293326">https://bugs.webkit.org/show_bug.cgi?id=293326</a>
<a href="https://rdar.apple.com/problem/151733273">rdar://problem/151733273</a>

Reviewed by Abrar Rahman Protyasha.

This completes the process begun in 293592@main.
This reduces the binary size as noted in 293804@main
293819@main was also preparation for this.

Legal approval and initial perf analysis links are included in the radar.
Initial perf analysis was somewhat inconclusive, so we&apos;ll be checking tomorrow.

Slight modifications from mpark::variant include these:
1. I include the license inline instead of in a separate file.
2. I include all needed headers inline instead of separate files.
3. I add a RELEASE_ASSERT and ignore -Wunsafe-buffer-usage around
the array access in mpark::lib::cpp14::array::operator[]
4. I unconditionally remove the definition of MPARK_EXCEPTIONS
because WebKit doesn&apos;t have C++ assertions enabled.
5. I include some auxiliary support in namespace WTF for things like
InPlaceType, VariantAlternative, VariantSize, etc.

* Source/WTF/wtf/Variant.h:
(mpark::lib::cpp14::array::operator[] const):
(mpark::lib::cpp17::detail::invoke_result&lt;void_t&lt;decltype):
(mpark::lib::cpp17::detail::has_addressof_impl::impl):
(mpark::lib::type_pack_element_impl::decltype):
(mpark::throw_bad_variant_access):
(mpark::detail::find_index):
(mpark::detail::find_index_impl):
(mpark::detail::trait):
(mpark::detail::common_trait):
(mpark::detail::common_trait_impl):
(mpark::detail::access::recursive_union::get_alt):
(mpark::detail::access::recursive_union::get_alt_impl::AUTO_REFREF_RETURN):
(mpark::detail::access::recursive_union::AUTO_REFREF_RETURN):
(mpark::detail::access::base::AUTO_REFREF_RETURN):
(mpark::detail::visitation::base::expected::but_got):
(mpark::detail::visitation::base::make_farray):
(mpark::detail::visitation::base::make_fmatrix_impl::impl):
(mpark::detail::visitation::base::make_fmatrix):
(mpark::detail::visitation::base::AUTO_RETURN):
(mpark::detail::visitation::base::make_fdiagonal):
(mpark::detail::visitation::variant::visitor::does_not_handle):
(mpark::detail::visitation::variant::AUTO_RETURN):
(mpark::detail::alt::alt):
(mpark::detail::MPARK_VARIANT_RECURSIVE_UNION):
(mpark::detail::base::base):
(mpark::detail::base::index_):
(mpark::detail::base::as_base):
(mpark::detail::base::data):
(mpark::detail::base::size):
(mpark::detail::MPARK_VARIANT_DESTRUCTOR):
(mpark::detail::constructor::ctor::operator() const):
(mpark::detail::constructor::construct_alt):
(mpark::detail::constructor::generic_construct):
(mpark::detail::MPARK_VARIANT_MOVE_CONSTRUCTOR):
(mpark::detail::MPARK_VARIANT_COPY_CONSTRUCTOR):
(mpark::detail::assignment::emplace):
(mpark::detail::assignment::assigner::operator() const):
(mpark::detail::assignment::assign_alt):
(mpark::detail::assignment::generic_assign):
(mpark::detail::MPARK_VARIANT_MOVE_ASSIGNMENT):
(mpark::detail::MPARK_VARIANT_COPY_ASSIGNMENT):
(mpark::detail::impl::assign):
(mpark::detail::impl::swap):
(mpark::detail::impl::swapper::operator() const):
(mpark::detail::impl::move_nothrow const):
(mpark::detail::is_non_narrowing_convertible::impl):
(mpark::variant::noexcept):
(mpark::variant::emplace):
(mpark::get):
(mpark::detail::convert_to_bool::operator() const):
(mpark::operator==):
(mpark::operator!=):
(mpark::operator&lt;):
(mpark::operator&gt;):
(mpark::operator&lt;=):
(mpark::operator&gt;=):
(mpark::detail::any):
(mpark::visit):
(mpark::detail::all_impl):
(mpark::detail::all):
(mpark::DECLTYPE_AUTO_RETURN):
(mpark::noexcept):
(std::get):
(WTF::visit):

Canonical link: <a href="https://commits.webkit.org/296355@main">https://commits.webkit.org/296355@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/1c78da2f82f8f2a544efd65f4dc98da004a9fe48

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/108289 "1456 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/27950 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/18372 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/113498 "Built successfully") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/58728 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/110252 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/28639 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/36499 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/82217 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/58728 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/111237 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/22701 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/97546 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/62653 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/22114 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/15684 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/58230 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/100853 "Built successfully and passed tests") | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/92069 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/15740 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/116620 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/106848 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/35343 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/26025 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/91250 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/35716 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/93823 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/91051 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/35934 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/13705 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/31100 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/17487 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/35245 "Built successfully") | [❌ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/40784 "Found 9 new failures in html/canvas/ImageBitmapRenderingContext.cpp, wasm/js/JSWebAssemblyStruct.h, wasm/js/JSWebAssemblyArray.h, wasm/js/JSWebAssemblyException.cpp, wasm/js/JSWebAssemblyStruct.cpp, workers/WorkerInspectorProxy.cpp") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/131134 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/34963 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/35593 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/38317 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/36643 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

aprotyas commented on Source/WTF/wtf/Variant.h:
```diff
@@ -25,9 +25,2919 @@
+
 #include <utility>
 #include <variant>
-
```
Nit: keep this line break?

achristensen07 committed:
Use mpark::variant on Cocoa platforms
https://bugs.webkit.org/show_bug.cgi?id=293326
rdar://problem/151733273

Reviewed by Abrar Rahman Protyasha.

This completes the process begun in 293592@main.
This reduces the binary size as noted in 293804@main
293819@main was also preparation for this.

Legal approval and initial perf analysis links are included in the radar.
Initial perf analysis was somewhat inconclusive, so we'll be checking tomorrow.

Slight modifications from mpark::variant include these:
1. I include the license inline instead of in a separate file.
2. I include all needed headers inline instead of separate files.
3. I add a RELEASE_ASSERT and ignore -Wunsafe-buffer-usage around
the array access in mpark::lib::cpp14::array::operator[]
4. I unconditionally remove the definition of MPARK_EXCEPTIONS
because WebKit doesn't have C++ assertions enabled.
5. I include some auxiliary support in namespace WTF for things like
InPlaceType, VariantAlternative, VariantSize, etc.

* Source/WTF/wtf/Variant.h:
(mpark::lib::cpp14::array::operator[] const):
(mpark::lib::cpp17::detail::invoke_result<void_t<decltype):
(mpark::lib::cpp17::detail::has_addressof_impl::impl):
(mpark::lib::type_pack_element_impl::decltype):
(mpark::throw_bad_variant_access):
(mpark::detail::find_index):
(mpark::detail::find_index_impl):
(mpark::detail::trait):
(mpark::detail::common_trait):
(mpark::detail::common_trait_impl):
(mpark::detail::access::recursive_union::get_alt):
(mpark::detail::access::recursive_union::get_alt_impl::AUTO_REFREF_RETURN):
(mpark::detail::access::recursive_union::AUTO_REFREF_RETURN):
(mpark::detail::access::base::AUTO_REFREF_RETURN):
(mpark::detail::visitation::base::expected::but_got):
(mpark::detail::visitation::base::make_farray):
(mpark::detail::visitation::base::make_fmatrix_impl::impl):
(mpark::detail::visitation::base::make_fmatrix):
(mpark::detail::visitation::base::AUTO_RETURN):
(mpark::detail::visitation::base::make_fdiagonal):
(mpark::detail::visitation::variant::visitor::does_not_handle):
(mpark::detail::visitation::variant::AUTO_RETURN):
(mpark::detail::alt::alt):
(mpark::detail::MPARK_VARIANT_RECURSIVE_UNION):
(mpark::detail::base::base):
(mpark::detail::base::index_):
(mpark::detail::base::as_base):
(mpark::detail::base::data):
(mpark::detail::base::size):
(mpark::detail::MPARK_VARIANT_DESTRUCTOR):
(mpark::detail::constructor::ctor::operator() const):
(mpark::detail::constructor::construct_alt):
(mpark::detail::constructor::generic_construct):
(mpark::detail::MPARK_VARIANT_MOVE_CONSTRUCTOR):
(mpark::detail::MPARK_VARIANT_COPY_CONSTRUCTOR):
(mpark::detail::assignment::emplace):
(mpark::detail::assignment::assigner::operator() const):
(mpark::detail::assignment::assign_alt):
(mpark::detail::assignment::generic_assign):
(mpark::detail::MPARK_VARIANT_MOVE_ASSIGNMENT):
(mpark::detail::MPARK_VARIANT_COPY_ASSIGNMENT):
(mpark::detail::impl::assign):
(mpark::detail::impl::swap):
(mpark::detail::impl::swapper::operator() const):
(mpark::detail::impl::move_nothrow const):
(mpark::detail::is_non_narrowing_convertible::impl):
(mpark::variant::noexcept):
(mpark::variant::emplace):
(mpark::get):
(mpark::detail::convert_to_bool::operator() const):
(mpark::operator==):
(mpark::operator!=):
(mpark::operator<):
(mpark::operator>):
(mpark::operator<=):
(mpark::operator>=):
(mpark::detail::any):
(mpark::visit):
(mpark::detail::all_impl):
(mpark::detail::all):
(mpark::DECLTYPE_AUTO_RETURN):
(mpark::noexcept):
(std::get):
(WTF::visit):

Canonical link: https://commits.webkit.org/296355@main
